### PR TITLE
Keeping up to date with Bricks and FNCAS + minor cleanup.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

**Comment by dkorolev** [comment]
Uses https://github.com/KnowSheet/fncas/pull/5.

Thanks!
Dima

**Comment by dkorolev** [comment]
And merged in https://github.com/KnowSheet/AgreeDisagreeDemo/pull/13.

@sompylasar : Could you please take a look? Max is busy for the next few days. Thanks!

File: demo.cc [hunk 1]
```diff
@@ -219,16 +223,16 @@ class Cruncher final {
...
+    std::function<void(Snapshot&)> function_with_snapshot;
     FunctionMQMessage() = delete;
-    explicit FunctionMQMessage(std::function<void(Box&)> f) : function_with_box(f) {}
+    explicit FunctionMQMessage(std::function<void(Snapshot&)> f) : function_with_snapshot(f) {}
   };
 
   struct HTTPRequestMQMessage : schema::Base {
     Request request;
-    std::function<void(Request, Box&)> http_function_with_box;
+    std::function<void(Request, Snapshot&)> http_function_with_box;
```
Comment (sompylasar) [inline demo.cc]: _with_snapshot?

File: demo.cc [hunk 2]
```diff
@@ -317,29 +319,29 @@ class Cruncher final {
...
+      snapshot_.engagement.AddAction(static_cast<double>(a.ms));
       TriggerVisualizationUpdate();
     }
 
-    inline void operator()(FunctionMQMessage& message) { message.function_with_box(box_); }
+    inline void operator()(FunctionMQMessage& message) { message.function_with_snapshot(snapshot_); }
 
     inline void operator()(HTTPRequestMQMessage& message) {
-      message.http_function_with_box(std::move(message.request), box_);
+      message.http_function_with_box(std::move(message.request), snapshot_);
```
Comment (sompylasar) [inline demo.cc]: _with_snapshot?

File: demo.cc [hunk 3]
```diff
@@ -493,33 +496,37 @@ class Cruncher final {
...
+            .OutputFormat("pngcairo");
+      } else {
+        return "";
+      }
     }
 
     void TriggerVisualizationUpdate() {
       visualization_.MutableUse([this](Visualization& visualization) {
-        // Make a copy the `box_` to work with.
+        // Make a copy the `snapshot_` to work with.
```
Comment (sompylasar) [inline demo.cc]: Make a copy __of__ the ...

File: demo.cc [hunk 4]
```diff
@@ -586,7 +591,7 @@ class MixpanelUploader final {
   MixpanelUploader(const std::string& demo_id, const std::string& mixpanel_token)
       : demo_id_(demo_id), mixpanel_token_(mixpanel_token) {}
 
-  inline bool Entry(std::unique_ptr<schema::Base>& entry, size_t index, size_t total) {
+  inline bool Entry(const std::unique_ptr<schema::Base>& entry, size_t index, size_t total) {
```
Comment (sompylasar) [inline demo.cc:594]: I had issues with const entry in mixpanel branch. Has this changed?

**Comment by sompylasar** [comment]
@dkorolev LGTM. Minor nits.

**Comment by dkorolev** [comment]
Fixed the minor things.

The `const` issue:
1. The entry should be passed as non-`const`. This is to allowing moving `unique_ptr`-s out of the caller thread into message queues for other threads. If the design enforces the passed in objects to be mutable, it's easy (just `std::move` them away), while if the design does not, it's up to the user to clone those objects, which is a pain for polymorphics.
2. A non-`const` entry can be passed on to a `const`-accepting method. So this part is solid.
3. Nonetheless, `map::operator[]` is mutable in C++ (it will create non-existing entries with default values, in order to be able to return references), and to keep using array-style indexing I simply kept one of two `Entry`-accepting method take a mutable input.

@sompylasar : PTAL. Thanks!

**Comment by sompylasar** [comment]
Thanks for the thorough explanation!

Which is preferred: having const or not in the Entry method signature? I
saw both around the code, and I've recently copy-pasted the non-const one
into the db/test.
25.03.2015 9:06 пользователь "Dima" notifications@github.com написал:

**Comment by dkorolev** [comment]
I'd suggest we go with non-const and `override` (`virtual bool Entry(SomeType& entry, size_t, size_t) override { ... }`) whenever possible.

Even though currently the dispatching is done via template magic, not virtual functions table, and thus `virtual` and `override` are not necessary, I can totally see a lot of usecases for them in the future, and would prefer to be proactively ready for them.

**Comment by sompylasar** [comment]
@dkorolev Maybe add this suggestion somewhere into the Sherlock docs?

**Comment by dkorolev** [comment]
:+1: https://github.com/KnowSheet/Sherlock/issues/11

**Pull request merged**